### PR TITLE
Rewrite contrib guidelines: shorter & friendlier

### PR DESCRIPTION
**Pull request by adriaanm**
Refactored to be less "just do it this way",
and more "here's how you can contribute more effectively".

Tried to motivate requirements, strike lighter tone in general.

Review by @ms-tg, @dickwall, and @som-snytt.

### CODE REVIEW

**Comment by adriaanm** [comment]
It's a complete rewrite, so I recommend [viewing the whole file](https://github.com/adriaanm/scala/blob/contrib/CONTRIBUTING.md)

**Comment by mpociecha** [comment]
Maybe could you add also this link: http://docs.scala-lang.org/scala/? Especially 'Tips and tricks' section is really helpful. This link is included in README.md but after adding it also here this document would aggregate probably all needed information.

By the way, this version seems to be way more informative and friendly to newcomers than the old one. :+1:

**Comment by Ichoran** [comment]
I think the bit about scala-internals is now a little too far the other way.  I think it's a good idea to encourage larger changes to be discussed before spending time implementing them.

**Comment by Kwestor** [comment]
Really nice Adrian!

What I think is missing here is a short instructions on how to setup the project, as this was problematic to me personally. Especially useful are instructions on how to build project and import to IDE of your choice. Building tips'n'tricks are described under link provided by @mpociecha.

**Comment by ms-tg** [comment]
I think the tone is better, but it is still _too long_. It says, **read all this text, and all the links, before even thinking of trying anything**.

I would put up front, a link to README.md inside the project. The README should actually be useful, and tell you the steps necessary to get setup in IntelliJ or Eclipse to work on and test the library first, and then the compiler second. 

There should be a shell script that loads all env variables and lets you run tests exactly how they'll be run inside CI -- this currently doesn't exist.

So for the document, the first section should be a one sentence: "We're so happy that you want to contribute to Scala! Please go ahead and [fork the repo](https://help.github.com/articles/fork-a-repo/), clone it locally, and follow [these easy steps](https://github.com/scala/scala/blob/2.11.x/README.md) to get setup to develop".

The following sections should be like what is written, but I would advise to make them shorter and more to the point?

**Comment by ms-tg** [comment]
Basically, what is the goal of this document? As it was originally written, I read the goal primarily one of _filtering_ -- these are the types of pull requests we _do not_ want -- and here's how to do it perfectly.

However, per Dick Wall, I believe it's time to recognize that what works better for other open source projects of a similar size is to adopt a goal of _enabling_ -- here's the minimal, brain-dead simple steps to be able to write a test and a fix for a perceived issue.

The correct time to advise someone to take an issue to the mailing list, for example, is _in response to a PR_! I think the main concern should be to get more people able and willing to build Scala and especially the standard library, and able to test and fix bugs in them. Filtering and guiding in to discussion, and sending a link on how to _improve_ your existing PR can come later, as it's done in many other projects.

Does that make sense to anyone?

**Comment by ms-tg** [comment]
@Ichoran: Does my point above about filtering vs enabling as primary goal of the reading experience make any sense to you? I see that you want more emphasis on pre-discussion in scala-internals, but I see that as sending the wrong message if the goal is actually to enable more people. You can guide discussion to the mailing list once you have the luxury of a submitted PR, right?

**Comment by dickwall** [comment]
Excellent re-write - I have been thinking after reading conversations over the last few days that we should maybe have something about the desire for communication on scala-internals as well prior to making changes (and by that I mean in general, before the specifics about documentation or code). It should not hold up this PR necessarily and the wording might take some massaging, but the points to hit are:
- Avoid duplication and conflicts
- Announces intent to others
- Invites comment and potentially help from others that may have tried
- Software development on a large code base works best when it is a somewhat social activity
- This step isn't there to slow you down - announce intent and get on with it, you are not asking permission, and silence implies consent anyway

Something like that, if not in the contributing guide, I intend to cover these points in the community reboot information, but it might still be desirable to touch on this in the contributing guide since I think we would all rather see this than not, right? So if we can explain that it is a potential benefit, and not a latch, that might help.

**Comment by adriaanm** [comment]
Thank you for the feedback everyone! I'll move the crucial info closer to the top (and import from README.md), and make it more concise overall.

**Comment by adriaanm** [comment]
One thing about filtering is that we don't want to waste people's time. It can be frustrating to spend a lot of time developing a PR that we can't accept (due to binary compatibility requirements between minor releases, for example).

**Comment by ms-tg** [comment]
@adriaanm "It can be frustrating to spend a lot of time developing a PR that we can't accept (due to binary compatibility requirements between minor releases, for example)." <-- so I just went through this experience. I still think it's better for the project that people submit a PR as they are motivated to, and if necessary on their first time you can then explain the binary compatibility policy and politely ask then to target the next major release.

Basically, it comes down to attracting/enabling vs filtering, and I think any filtering action is premature here. You don't go to a first date or a job interview, or even worse email ahead of time, with a list of every filter criteria that you want to apply. You try to attract / enable the best opportunities, and as the relationship develops you ask the other party to consider more and more of your filter criteria. Or maybe that's just me?

**Comment by adriaanm** [comment]
Good point. I agree with you -- maybe we should self-select contributors who are willing to go through the PR process without a guarantee of success (thank you!). We have had the opposite as well ("What should I do for my PR to be merged with certainty?"), but maybe we shouldn't worry about those, and instead worry about putting the enthusiasts off with too many caveats and requirements.

**Comment by ms-tg** [comment]
Yeah, I mean YMMV but that's my point of view. The criteria and the caveats etc still exist of course -- it's just not blaring them up front may reduce the friction and end up with more people to filter at the end ;)

**Comment by adriaanm** [comment]
Ok, I think I've implemented everyone's feedback, so I'm going to merge.

**Comment by adriaanm** [comment]
Thank you, @mpociecha, @Kwestor and @ms-tg!
